CPT / Site Settings: Add options for toggling custom content types

aduth opened the pull request:
Closes #4481

This pull request seeks to add a new fieldset to the site settings writing form for toggling whether custom post types are enabled on a site. As described in #4481, until now a user must have navigated to wp-admin to manage these settings.

![CPT Settings](https://cloud.githubusercontent.com/assets/1779930/14252123/efe5a752-fa53-11e5-89e7-8f3ef09110ef.png)

__Implementation notes:__

- Many themes add default support for custom post types. Support for this behavior does not (but could) exist at the REST API. Instead, the checkboxes are shown as checked and disabled when the the setting is not enabled but the post type exists for the site (implying that the theme adds support).
 - ![Implicitly activated](https://cloud.githubusercontent.com/assets/1779930/14252205/40319306-fa54-11e5-829e-062daa97dd6a.png)
- Settings fields for managing custom post types were only recently added to the REST API (r133680-wpcom). Therefore, Jetpack support will not be available until these changes have been migrated to the plugin. The fieldset is hidden for Jetpack sites not satisfying a version compare for `>=4.0.0`.

__Testing instructions:__

Verify that the fields are checked when enabled for a site, are disabled when default support exists for a theme, and can be toggled and saved. After a save, ensure that the corresponding sidebar navigation item is added or removed based on your change. Verify that the form fields are not shown for Jetpack sites, and that navigating directly between sites (via site picker) updates the field values correctly.

1. Navigate to the [writing settings page](http://calypso.localhost:3000/settings/writing)
2. Select a site
3. Note that...
 - If a Jetpack site, the custom post types fieldset is not shown
 - Otherwise, a custom post types fieldset is shown
4. Note that the current checked status reflects the settings for the site
5. Change the checked status
6. Save the form
7. Note that the sidebar item is removed or added based on your chnage in step 5
8. Refresh the page
9. Note that the current status still reflects the settings for the site

__Caveats:__

- Users cannot (yet) manage "display at most" post count setting. Some design feedback would be appreciated here. The setting does exist at the REST API and could be added in this or a subsequent pull request.
- Jetpack support is not yet available

timmyc commented on client/my-sites/site-settings/custom-post-types-fieldset/index.jsx:
```diff
@@ -0,0 +1,143 @@
+				{ siteId && (
+					<QueryPostTypes siteId={ siteId } />
+				) }
+				<FormLabel>{ translate( 'Custom Post Types' ) }</FormLabel>
```
It looks like the text here is form wp-admin, where the section is titled "Custom Content Types" - as does the linked documentation http://jetpack.com/support/custom-content-types/.  Should we retain that nomenclature here?

aduth replied:
>Should we retain that nomenclature here?

Yeah, I think that'd be appropriate.

aduth replied:
>Should we retain that nomenclature here?

Updated in 5041fb9.

timmyc commented on client/my-sites/site-settings/custom-post-types-fieldset/index.jsx:
```diff
@@ -0,0 +1,143 @@
+	requestingPostTypes: PropTypes.bool,
+	requestingSettings: PropTypes.bool,
+	settings: PropTypes.object,
+	recordEvent: PropTypes.func,
```
From line #79 above do we also need `className` and `value` here as propTypes?

aduth replied:
>do we also need `className` and `value` here as propTypes?

Indeed!

aduth replied:
>do we also need `className` and `value` here as propTypes?

Updated in c4c28d1.

timmyc commented:
Not sure if I read this part of the comments above correctly or not:

```
Many themes add default support for custom post types. Support for this behavior does not (but could) exist at the REST API. Instead, the checkboxes are shown as checked and disabled when the the setting is not enabled but the post type exists for the site (implying that the theme adds support)
```

But on a site whose theme supports testimonials I am seeing the following:

<img width="918" alt="site_settings_ _trout_bummin_ _wordpress_com" src="https://cloud.githubusercontent.com/assets/22080/14263045/5ac16296-fa6d-11e5-9479-cecbf47d1917.png">

vs wp-admin:

<img width="645" alt="writing_settings_ _trout_bummin_ _wordpress" src="https://cloud.githubusercontent.com/assets/22080/14263055/67bdf5fe-fa6d-11e5-97da-238730ae15ee.png">

aduth commented:
>But on a site whose theme supports testimonials I am seeing the following:

I've seen this before myself, and I assume it can occur when you had enabled the option previously on a theme where default theme support did not exist. You can confirm this in the [console](https://developer.wordpress.com/docs/api/console/) by checking whether the response property for `jetpack_testimonial` is `true` for the `GET /sites/%s/settings` endpoint. It's a bit of an ugly consequence of the lack of proper support for theme support detection here, though admittedly not entirely misleading.

I'm also a bit curious about the Testimonials label in the sidebar from your screenshot. Why's it showing in French while other labels are in English? Result of stale persisted state? Did it eventually refresh itself?

timmyc commented:
> I'm also a bit curious about the Testimonials label in the sidebar from your screenshot. Why's it showing in French while other labels are in English? Result of stale persisted state? Did it eventually refresh itself?

Yeah I had noticed that too.  I had previously been testing your branch with instructions for testing in French.  It is gone as of this morning, but I had changed branches + done a hard refresh and it persisted.

timmyc commented:
Very strange results from the API for my theme.  Guessing out of scope for this PR but, if I disable Portfolio, the API then returns:

![wordpress_com_console_](https://cloud.githubusercontent.com/assets/22080/14296392/c51ef4e8-fb2e-11e5-9af9-a5e76ad88991.png)

Which is rendered in Calypso as:

![site_settings_ _trout_bummin_ _wordpress_com_](https://cloud.githubusercontent.com/assets/22080/14296421/dd334fb6-fb2e-11e5-9ebb-cfae16a8210d.png)

If I re-enable Portfolio, I then see a correct representation in Calypso:

![site_settings_ _trout_bummin_ _wordpress_com_](https://cloud.githubusercontent.com/assets/22080/14296453/131d4460-fb2f-11e5-83e9-f35ae054ee79.png)

Then I can uncheck Portfolio, and save.  At that point I can verify that Portfolio is unchecked in wp-admin and calypso.  Once I do a hard refresh on the page, I am back to the initial screenshot where the checkbox inputs for both CPTs are disabled.

I experience the same issues with the Port theme which also enables testimonials by default.

I did verify that Jetpack is working as expected though.

aduth commented:
I'm starting to think we might need proper API support for indicating theme CPT support :laughing:

timmyc commented:
Yes we might.  It just is confusing when these different states can be reached and do not reflect what is shown in wp-admin.

mtias commented:
I suggest we use the toggle component instead of checkboxes.

aduth commented:
Coming back to this one. I've rebased to resolve merge conflicts and made the following changes:

- Use `<FormToggle />` instead of checkboxes for toggling active content types
 - Required some changes to `<FormToggle />` to (a) look better when a label child is included (not used elsewhere) and (b) call `onChange` when label text is clicked (previous implementation did not work as expected)
- Hide form options altogether for Jetpack sites. I've noted that the endpoint changes for `GET /sites/%s/settings` have made their way to Jetpack master, but I'd be more comfortable holding off for more thorough testing once it at least reaches Jetpack beta.

On the point of inferring theme support, I have not made any changes to retrieve this from the API. It's not straight-forward where we'd include this information (there's three candidates I have in mind, each with a catch). I'm rather content with the current behavior, in that the only awkward situation is if a user changes themes from one which had not previously supported a custom content type but where the user had explicitly enabled it to a theme where that type is supported, the option will show as checked, and if the user unchecks it and saves, it'll then show the "Theme supports" label. Not only does this feel pretty edgy, but at worst it communicates "hey, while you wanted to disable it for whatever reason, it's supported by default for your theme". @timmyc I'll follow-up with you directly on the API limitations, but let me know whether you think this will continue to be problematic as implemented here.

timmyc commented:
Toggles look really nice - good job.

I am experiencing this when refreshing the page though:

![custom-content-types](https://cloud.githubusercontent.com/assets/22080/14992877/125ee1a6-111d-11e6-8a56-0b4c84bbbb81.gif)

Is this the edge case you spoke of in your last comment?

aduth commented:
>Is this the edge case you spoke of in your last comment?

Hmm, no. This may be a regression of a condition I had removed that had been checking that post types or settings were loading. I'll want to take another look at this. I'd removed it since it was causing the toggles to flicker briefly as not enabled after saving.

aduth commented:
>I am experiencing this when refreshing the page though:

@timmyc It turned out to be a simple enough fix. When considering whether a post type is enabled by the theme, we should check that the site setting is explicitly false, not that it's falsey, which will pass the check if the settings are still loading. See rebased 3a092d3.

timmyc commented:
👍  that fixed up the issue noted above.  Tested across a variety of configurations.  The only thing that seems a bit off to me now is how the toggles are disabled after the save takes place.  Though attempts to interact with other form controls after the save happens - they all are disabled as well.

Just figured I'd note it here, otherwise LGTM

![toggle-cpt](https://cloud.githubusercontent.com/assets/22080/15193811/93cd0b52-1774-11e6-9f74-527aeec916b4.gif)

mtias commented:
Perhaps the disabled state should be light blue for the one in Testimonials.

mtias commented:
@aduth @timmyc can we include a small description on what these things are below each toggle? It would also improve the design a bit.

aduth commented:
>Perhaps the disabled state should be light blue for the one in Testimonials.

Assuming you'd recommend this to be the case for any checked but disabled `<FormToggle />` ?

>can we include a small description on what these things are below each toggle?

Sure, I'll plan to add these.

I also think I might hide this setting behind the `manage/custom-post-types` feature flag, since it may be detrimental to the user experience to allow them to toggle the custom types, but only manage them through the wp-admin interface.

mtias commented:
> Assuming you'd recommend this to be the case for any checked but disabled <FormToggle /> ?

Yes, otherwise it briefly seems like the setting itself is being toggled to me.

mtias commented:
> I also think I might hide this setting behind the manage/custom-post-types feature flag

Sounds good.

aduth commented:
Rebased, squashed, and pushed revisions to address last few points:

- Display disabled, checked `<FormToggle />` with faded blue background (6d3d048)
- Add descriptions for custom content types (4f989ba), sourced from [Jetpack support page](https://jetpack.com/support/custom-content-types/)
- Hide settings behind CPT feature flag (ac43cac)

__New settings descriptions:__

![image](https://cloud.githubusercontent.com/assets/1779930/15295207/754ff760-1b5d-11e6-9907-28c4f93ad895.png)

Could use some design love, though consider that we're using `<FormFieldExplanation />` already for indicating default theme support, hence the use of unclassed `<p>` tags.

__New `<FormToggle />`:__

Before|After
---|---
![Before](https://cloud.githubusercontent.com/assets/1779930/15295167/3a6d8efa-1b5d-11e6-8d2f-fdcfc696d96d.png)|![After](https://cloud.githubusercontent.com/assets/1779930/15295091/db49da3c-1b5c-11e6-88cb-d11c8bc0a724.png)

_([Example in DevDocs Design](http://calypso.localhost:3000/devdocs/design/form-fields))_

timmyc commented:
New changes look great - the new disabled state on the form toggle is a great touch.  All looks well when disabling the feature flag too.

My only small nit with this would be the length of the explanations.  It is fine to go as-is, but when viewing this settings screen on a smaller viewport, these paragraphs, along with the PressThis text create for a massive wall of text.

Granted these screens probably aren't visited often, just thought I would mention it.

🚢

aduth commented:
@mtias @timmyc We could try editor-style info bubbles?

![info bubble](https://cloud.githubusercontent.com/assets/1779930/15393854/7b8d6ae4-1d9d-11e6-8ca4-81f3a13855ae.png)

I may still move forward with merging anyways, since as you mentioned, we can iterate on this in subsequent pull requests.

timmyc commented:
> I may still move forward with merging anyways, since as you mentioned, we can iterate on this in subsequent pull requests.

Yes I say merge this in now and we can revisit later if needed.  I did think info bubble might be worth exploring too.

mtias commented:
In terms of design I think we should replicate how this WordAds card looks:

https://cloud.githubusercontent.com/assets/3775068/15296615/651bc0d4-1b96-11e6-9f21-93b8a34d50ba.png

For each CPT.

rickybanister commented:
I agree with @mtias on this, we don't put the toggles to the left of the labels often (if anywhere) in calypso. 

The WordAds example is very clear, makes room for the description, and helps break up the monotony of settings. Just make sure to use the `sectionHeader` treatment instead of that one-off bold gray-dark thing from WordAds.

cc @hugobaeta

aduth commented:
Thanks @mtias and @rickybanister . I've pushed updates to make use of `<SectionHeader />` and `<Card />` for displaying the settings. How does this look to you?

![image](https://cloud.githubusercontent.com/assets/1779930/15402471/5a4e0c74-1dc2-11e6-85a5-ff34e98130aa.png)

We could also use `<FoldableCard />` instead to hide the descriptions behind an expander, but I think it's reasonable to make the descriptions available by default.

rickybanister commented:
That looks great to me. I think using foldable card might make it even more complicated with the toggle next to the open/close arrow. 

It's good to see these descriptions if it's the first time you're seeing these settings.

aduth commented:
Thanks for the feedback @rickybanister ! I'm going to go ahead and merge this now.

folletto commented:
Great to see this in! 👍
